File: ios/Shared/QueryInput.swift
import Foundation

/// Boundary validation for user text crossing into the native engine. The embedding model only
/// reads ~256 word-pieces, so anything past the cap is dead weight in the FFI marshal.
enum QueryInput {
    static let maxLength = 500

    /// Trimmed, length-capped query — or nil if there's nothing to send.
    static func sanitize(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        return String(trimmed.prefix(maxLength))
    }
}

File: android/app/src/main/kotlin/io/dotnetnativeinterop/model/QueryInput.kt
package io.dotnetnativeinterop.model

/**
 * Boundary validation for user text crossing into the native engine, mirrored from iOS
 * (`ios/Shared/QueryInput.swift`). The embedding model only reads ~256 word-pieces, so anything
 * past the cap is dead weight in the JNI marshal.
 */
public object QueryInput {
    public const val MAX_LENGTH: Int = 2000

    /** Trimmed, length-capped query — or null if there's nothing to send. The cap bounds the JNI
     *  marshal without cutting below the model's ~256-word-piece window; truncation never splits
     *  a surrogate pair (an unpaired surrogate would corrupt the UTF-8 marshal). */
    public fun sanitize(raw: String): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        var capped = trimmed.take(MAX_LENGTH)
        if (capped.isNotEmpty() && capped.last().isHighSurrogate()) capped = capped.dropLast(1)
        return capped
    }
}
